### DISCUSSION TITLE
Search groups by their attributes

### DISCUSSION THREAD

**Discussion by alice-wondered** [keycloak/keycloak#12997] [Feedback]
Looking for feedback on an enhancement to searching for groups via their attributes as described by this ticket: https://github.com/keycloak/keycloak/issues/12964

**Comment by jbman** [comment]
We would welcome this feature, too. In one of the Keycloak systems we operate, we create a group per customer. We need to include the external identifier (customer number) into the group name, so that the group can be found by name. Would help if the group can also be found by other criteria.

Related topic:
I'm missing sections in the documentation which explain how queries for users or clients work. The only thing stated for users or clients is the URL parameter `q`, but without any explanation or example. (https://www.keycloak.org/docs-api/18.0/rest-api/#_clients_resource)

When looking at the test there are different query formats are supported:
`key1:val1 nokey key2:"val 2" key3:val3`
https://github.com/keycloak/keycloak/blob/main/services/src/test/java/org/keycloak/utils/SearchQueryUtilsTest.java

So it's even possible to search for an attribute value without key.

**Reply by hmlnarik** [reply]
> So it's even possible to search for an attribute value without key.

In the context of users, this is full-text search for specific fields (username, first / last name, email); in the context of clients the behaviourr is undefined. It should not be confused with search for arbitrary value in all user attributes though.

**Comment by vramik** [comment]
@Redhat-Alice Thank you for starting the discussion. I think it makes sense to add such a capability to groups. I'm happy to help with with JPA map storage implementation.

@jbman Also I agree that current documentation regarding similar feature for clients is not great and it'd be worth to improve it.

**Comment by jbman** [comment]
I have listed some specific behavior for the current attribute search at the users endpoint. Such information would be helpful for developers at the Keycloak Admin API documentation. 

## Search for the exact value

By default the search finds all entries, where the attribute value contains the search string. The search `q=firstName:"Ali"` will find an entry who has first name "Ali" and also one who has first name "Alice" or first name "Dr. Ali".

If you want to find exact values only, add the query parameter `exact=true`, e.g. `q=firstName:"Ali"&exact="true"`. Warning: Even with the exact=true parameter the search is not case sensitive. An exact search for "Ali" will also return an entry with "ali".

## Search for a value with whitespace

You can surround the value with parentheses, e.g. `q=fullName:"Dr. Strange"`. You need to do this, if your search criteria value contains a whitespace,

## Search for multiple criteria (AND)

You can search for multiple key-value pairs, e.g. `q=postalCode:12345 businessCustomer:false`. This will find all users where both criteria match (AND semantics). If other dedicated query parameters like firstName or lastName are also used these are added with AND semantics. too.

## Check if user exists

If you just want to check if a user exists, add query parameters `&max=1&briefRepresentation=true` to limit the result set to one entry and minimize the size of this entry.

**Reply by tsavinash** [reply]
Is it possible to achieve OR semantics . Suppose q=postalCode:12345 OR businessCustomer:false our expectation is to get all the records which are matching either one of the condition . I appreciate if you can help if any other way to achieve this .
